### DISCUSSION TITLE
Installation: wait until Solr server is running

### DISCUSSION THREAD

**Discussion by anna-uol** [OPUS4/community#26] [Q&A]
Liebes OPUS-Team,

Um zu evaluieren, ob wir OPUS als neues Repository an unserer Bibliothek verwenden ist mein Ziel eine Test-Installation von OPUS 4. 
Unglücklicherweise endet der Durchlauf des Installationsskriptes mit dem oben genannten echo: 

wait until Solr server is running.........................

Solr ist ordnungsgemäß installiert und aktiv.
Ich erhalte keine Fehlermeldung. 
Hier einige Spezifikationen: 
Ich arbeite auf einer VM Kubuntu 22, php 8.1.2, solr 7.7.3

Mir ist nicht klar, worin das Problem besteht und würde mich über alle Ideen und Hinweise freuen, die mich zu einer erfolgreichen Installation bringen. 
![endstand](https://github.com/OPUS4/community/assets/149086171/76268957-6a0c-48d6-b233-77de5d2a4db7)

**Comment by j3nsch** [comment]
Es schwer zu sagen, aus der Ferne, woran das liegen könnte. Wir haben auch leider keine automatischen Tests für die Installationsskripte, so dass ich gar nicht sagen kann, ob alles wirklich funktionieren sollte, ohne selber noch mal manuelle Tests durchzuführen. Die Skripte werden wenig verwendet. Die meisten Instanzen befinden sich im Hosting und dort werden eigene Methoden verwendet für das Aufsetzen neuer Instanzen. Und für produktive Systeme sollte auch manuell Hand angelegt werden.  

So, aber jetzt zu den Dingen, die hoffentlich helfen. Nach dem Punkt an dem die Installation auf Solr wartet, passiert nicht mehr viel. Es wird die Indexierung der Testdaten durchgeführt und dann wahlweise der Apache2 neu gestartet.  Das Skript kann also abgebrochen werden und dann diese beiden Schritte manuell ausgeführt werden. 

Im Verzeichnis der OPUS 4 Installation:

    bin/opus4 index:index   
    sudo service apache2 restart
    
Man kann die Datenbank und den Index später auch jederzeit wieder zurück setzen und mit den Testdaten befüllen:

    ant reset-testdata
    
Dabei zeigt sich dann auch gut, ob die Datenbank und die Verbindung zu Solr funktionieren.

Es gibt häufig Schwierigkeiten mit den Permissions für die Verzeichnisse, da insbesondere die Log-Dateien manchmal von den Skripten, manchmal vom Apache2 angelegt werden. Das kann mit dem folgenden Skript gefixt werden, damit der lokale User und Apache2 Zugriff auf die Dateien haben.

    sudo bin/set-file-permissions.sh
         
Ob Solr läuft, kann man gut prüfen, indem man die Admin-Seiten von Solr aufruft. Die sind häufig unter folgender Adresse zu finden:

    http://localhost:8983

Dort kann dann auch geprüft werden, ob ein Solr-Core für OPUS 4 angelegt wurde. Die Konfiguration für die Solr-Verbindung findet sich 
in 

    application/configs/config.ini
    
Vielleicht ist dabei etwas schief gelaufen und die Verbindungsdaten stimmen nicht.     

Eine Alternative zu dieser Art der Installation für Testzwecke, ist die Verwendung von Vagrant. Die Anforderungen dafür sind ziemlich niedrig und es macht das Erzeugen einer Testinstanz sehr einfach. Es gibt dazu eine kurze Anleitung.

https://github.com/OPUS4/application/wiki/Vagrant

Vagrant funktioniert mit Linux, Windows, Mac OS-X und ist ein Tool, um automatische eine VM zu erzeugen,  in der dann eine vollständige Testinstanz läuft. Die Schritte dafür sind im `Vagrantfile` definiert. Die Datei stellt damit auch eine Referenz für das manuelle Einrichten einer Test-Instanz dar. Vor allem kann man mit Vagrant aber auch schnell mal eine Instanz für einen anderen Branch der OPUS 4 Entwicklung erzeugen, z.B. die nächste Version auf dem Branch **v4.8.1**, die dann Solr 9 verwendet.

Für produktive Systeme muss der Server ausreichend abgesichert werden und Apache Solr sollte nach der Anleitung in der offiziellen Dokumentation installiert werden. Das OPUS 4 Install-Skript ist wirklich nur für Testumgebungen gedacht. Für mehr Aufwand stehen nicht genug Entwickler zur Verfügung und wie gesagt, die Skripte werden kaum verwendet. 

Vielleicht finde ich die Zeit auch noch mal selber die Skripte zu testen, aber ich hoffe die Hinweise oben helfen Ihnen erst einmal weiter. 

Die Solr-Installation in den Skripten war immer etwas schwierig und wird in Zukunft wahrscheinlich komplett entfernt werden. Genau wie Apache2 und MySQL ist Solr einfach eine Voraussetzung für den Betrieb von OPUS 4. Die offiziellen Anleitungen und Tools für Solr sind gut. Der Aufwand das noch mal für OPUS 4 zu verpacken ist nicht wirklich gerechtfertigt und führt häufig nur zu Verwirrungen und Problemen.

**Comment by j3nsch** [comment]
Ich habe auch gerade noch mal das `Vagrantfile` für den aktuellen Release, **master** Branch, getestet, zur Sicherheit, unter Linux. Das funktioniert. Beim ersten Aufbauen der VM mit `vagrant up`, dauert es unter Umständen eine Weile (10+ Minuten). Das Vagrantfile für den **v4.8.1** Branch funktioniert auch. Manchmal verhält sich Vagrant mit Windows als Host etwas anders, aber wir haben alle bekannten Probleme gefixt. Bei Mac OS-X sind mir keine Probleme bekannt.
